### PR TITLE
Tests for #1507: Import MySQL column/table descriptions during database:reverse & added PDODataFetcher::fetchAll()

### PR DESCRIPTION
**Pull request by mringler**
Added tests for #1507, which adds an import of column and table descriptions during `database:reverse` on a MySQL database.

Also, while writing the tests, I again missed `PDODataFetcher::fetchAll()` (the class has a `fetch()` method, and in general just wraps PDO methods), so I added it in its own commit (didn't want to put it into its own PR, as I need it in the tests).

If merged, #1507 can be closed.

### CODE REVIEW

**Comment by mringler** [comment]
@dereuromark something's wrong with CI, tests remain in queued state forever. Started yesterday, between updating PRs. I think I saw them sitll running after the new merges, which shouldn't compromise CI anyway.
Can you have a look at that? Maybe you can restart them manually or something?

**Comment by dereuromark** [comment]
Looks like I also do not have enough rights to see what is going on. 
Someone with actual maintainer access needs to look into settings and CI topic.
@propelorm/propel2 @marcj Are you able to check this please?

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (gechetspr) [inline src/Propel/Generator/Reverse/MysqlSchemaParser.php:312]: could you add return types in new methods?

Also I would prefer if we can return Table class here in order to make modifying of this class more visible, but I can see that other methods in this class don't do this, so this is not mandatory

Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: Right, yes. Return type was missing in #1507, but I should have added it before. Thank you for finding.
Also added return type and missing parameters to `PDODataFetcher::fetchAll()`.

I have also updated the structure of the `addDescription()` methods. Let me know if this is more too your liking.

File: tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php
Comment (dmytro-dymarchuk) [inline tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php:63]: Small suggestion. To make code more clear better to not use abbreviations, so I suggest rename `$t` to `$table`.

File: tests/Propel/Tests/Generator/Reverse/AbstractSchemaParserTest.php
Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/AbstractSchemaParserTest.php:29]: This is the test class, it does not implement an interface. Or do you mean something else?

File: tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php
Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php:44]: I know. I left it on purpose to highlight that there is more going on in this test class.

File: tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php
Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php:52]: The `test...` functions generally do not have return types in this project. I think that is because some versions of phpunit are picky when it comes to that.

File: tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php
Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php]: sure

File: tests/Propel/Tests/Generator/Reverse/AbstractSchemaParserTest.php
Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/AbstractSchemaParserTest.php]: Sure, good idea

File: tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php
Comment (dmytro-dymarchuk) [reply tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php:44]: I still think it is not needed. We always know that `setUp` executed and test classes always contain something different from `setUp`.

Comment (mringler) [reply tests/Propel/Tests/Generator/Reverse/MysqlSchemaParserTest.php]: Ok, removed it

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (dereuromark) [inline src/Propel/Generator/Reverse/MysqlSchemaParser.php:342]: The var looks wrong, there is a different Syntax for such cases afaik.

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (dereuromark) [inline src/Propel/Generator/Reverse/MysqlSchemaParser.php:95]: This comment is usually noise

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php:342]: I looked [here](https://psalm.dev/docs/annotating_code/supported_annotations/), but could not find a dedicated annotation for this scenario.
The only other solutions I found was to silence the psalm error in the baseline file, which I wanted to avoid. Or add a clunky type cast, which ultimately casts a string to a string. Or store the return value in an otherwise useless variable, just for the var statement.
I was surprised that var works here, and agree that it looks a bit off. But next to the other options, I found it the least bad. Let me know if you prefer another option, or know another way to do it.

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: Agreed, removed it

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (dereuromark) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php:340]: You are right, without the $variable here this seems to be the syntax for returning.
All good, sry :)

**Review by dereuromark — APPROVED** [review]

**Comment by dereuromark** [comment]
One more reviewer needed

**Review by dmytro-dymarchuk — APPROVED** [review]

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (gechetspr) [inline src/Propel/Generator/Reverse/MysqlSchemaParser.php:313]: I would move this to the separate method

**Review by gechetspr — CHANGES_REQUESTED** [review]

**Comment by marcj** [comment]
@dereuromark admin is @PhilinTv.

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php:133]: Hmm, let me think about that.

As you already said, this breaks the style of how composition is done in the method. Which I don't even find all that confusing. But even if you do, adding inconsistencies definitely makes it worse. If this change is dear to your heart, I think you need to refactor all the `add...` methods in another PR.

Though I think I even like it better the way it is. Returning an object usually means one of two things: either it is a newly constructed object (the way list functions like map and reduce return a new array), or it is `$this`, so you can chain on it. Neither is true in this case, so it just looks like misunderstood pass-by-reference/pointer. Even worse, if I read something where the same variable gets reassigned several times, I would assume the author was just too lazy to introduce new variables. But if you assign to new variables here (what you explicitly encourage with your interface), they all point to the same object, and the whole thing is just a mess.
To me, it just adds noise, both on a structural level, and literally with just more ascii for what basically are noops. It really seems like a somewhat arbitrary change to the interface, that is supposed to mean something else.

I would argue, that whenever you pass an object, you implicitly pass permission to perform any change the object allows you to make through its interface. But if you want to make it clear to the reader that a function will change the passed object, there are several ways to do it that communicate it more cleanly and clearly:
- Simplest way is through appropriate function names, i.e. `addDescriptionToTable()` or `setTableDescription()` or such.
- You can just not pass on the object, and apply all changes locally: 
```php
$tableDescription = $this->loadTableDescription($table->getName());
$table->setDescription($tableDescription);
```
(This however would hurt readability of the composition function here, which is why I am doing this in another method.)
- You can introduce table builder objects, which is probably the most OOP approach. Then you can just tell the builder to load the description. Most clear, most elegant, most versatile - and probably about three days of work, rewriting the whole class. Not my focus.
- If you really want to use the interface, you can pass the object by alias, i.e. doing `addDescriptions: &$table -> void`. That way, it is clear that the function will do whatever it wants to the object, up to completely replacing it. However, we don't want to replace it, we want to update it. And lots of people will be even more confused.

Thinking about it, I am now convinced that the `addDescriptions()` function should not return the table object. I'll remove that again, and instead suffix the method with "ToTable".
But please, let me know your reasoning. This is obviously important to you, so I am curious about your approach to the issue. I am happy to change my mind.

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php:324]: Why though? Doesn't seem to make a big difference.
Changed it anyway. Happiness achieved, I hope.

Comment (gechetspr) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php:324]: Yep, thank you!

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (gechetspr) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: It's not a big deal actually. I agree that it's better to keep everything to the same style.

I like this approach, because IMO it makes code more descriptive and easy to read.
I don't need to check method body to understand that method returns modified object.

But I guess this is matter of taste and conventions :)

**Review by gechetspr — APPROVED** [review]

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: I think I used to do it that way sometimes. Thanks to the discussion, I am now convinced it is actually wrong, and I would request changes in a review. Funny how that goes. Well, guess I learned something today.

Comment (marcj) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: This `ToTable` affix is inconsistent with all the other reverse classes. I'd recommend to name it as before (and in all other classes) and not introduce a new naming style.

Comment (mringler) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: Good point, but too late, it was already merged. Probably needs a new PR, if you feel strongly about it.

**Comment by mringler** [comment]
@dereuromark if you want, you can close #1507 now

File: src/Propel/Generator/Reverse/MysqlSchemaParser.php
Comment (dereuromark) [reply src/Propel/Generator/Reverse/MysqlSchemaParser.php]: You can make a new PR here.

**Pull request merged**
